src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java

```diff
@@ -265,11 +265,43 @@ public long transferTo(OutputStream out) throws IOException {
             return transfer(rbc, fc);
         }
 
+        // ReableByteChannel -> WritableByteChannel
+        if (out instanceof ChannelOutputStream cos) {
+            ReadableByteChannel rbc = ch;
+            WritableByteChannel wbc = cos.channel();
+
+            if (rbc instanceof SelectableChannel rsc) {
+                synchronized (rsc.blockingLock()) {
+                    if (!rsc.isBlocking())
+                        throw new IllegalBlockingModeException();
+                    if (wbc instanceof SelectableChannel wsc) {
+                        synchronized (wsc.blockingLock()) {
+                            if (!wsc.isBlocking())
+                                throw new IllegalBlockingModeException();
+                            return transfer(rbc, wbc);
```

Comment:
This introduces the potential for deadlock as tock ordering is not specified for the transferXXX methods. Note that this same issue is problematic for some of the optimisations at the lower level in FileChannel too. We'll need to think about this topic.

Reply:
IIUC the JRE already has the same problem in *other* locations, so wouldn't it make sense to merge the current PR as-is (to gain its benefit already in the upcoming LTS release) and postpone the discussion about possible deadlocks into a separate PR which then covers *all* those existing problematic code locations as a whole? Because I do not see why to withhold *just this* PR due to a problem existing *already* in the JRE.

Reply:
> IIUC the JRE already has the same problem in _other_ locations, so wouldn't it make sense to merge the current PR as-is (to gain its benefit already in the upcoming LTS release) and postpone the discussion about possible deadlocks into a separate PR which then covers _all_ those existing problematic code locations as a whole? Because I do not see why to withhold _just this_ PR due to a problem existing _already_ in the JRE.

Indeed, the issues of async close of the source/target during transfer, and the potential for deadlock with transfer between bidirectional streams is mostly a lower level concern for methods like FileChannel.transferXXX. It may be that we have to specify an ordering that higher level participants (ChannelInputStream.transferTo) would need to follow.

Reply:
I do agree.